File: 算法/代码随想录算法PDF/00调试源代码/738.单调递增的数字.cpp
/*
 * @lc app=leetcode.cn id=738 lang=cpp
 *
 * [738] 单调递增的数字
 */

// @lc code=start
class Solution {
public:
    int monotoneIncreasingDigits(int n) {
        string s = to_string(n);
        int nineIndex = s.size();  // 变为9的最左坐标
        for (int i = s.size() - 1; i > 0; i--) {
            if (s[i - 1] > s[i]) {
                s[i - 1]--;
                nineIndex = i;
            }
        }
        for (int i = nineIndex; i < s.size(); i++) {
            s[i] = '9';
        }
        return stoi(s);
    }
};
// @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/738.单调递增的数字.py
#
# @lc app=leetcode.cn id=738 lang=python3
#
# [738] 单调递增的数字
#

# @lc code=start
class Solution:
    def monotoneIncreasingDigits(self, n: int) -> int:
        s = list(str(n))
        flag = len(s)
        for i in range(len(s) - 1, 0, -1):
            if int(s[i - 1]) > int(s[i]):
                s[i - 1] = str(int(s[i - 1]) - 1)
                flag = i
        for i in range(flag, len(s)):
            s[i] = '9'
        return int("".join(s))
# @lc code=end

